cpp/include/raft/util/bitset.cuh

```diff
@@ -0,0 +1,202 @@
+ * It provides a test() device function to check if a given index is set in the bitset.
+ *
+ * @tparam IdxT Indexing type used. Default is uint32_t.
+ */
+template <typename IdxT = uint32_t>
+struct bitset_view {
+  using BitsetT            = uint32_t;
+  IdxT bitset_element_size = sizeof(BitsetT) * 8;
+
+  _RAFT_HOST_DEVICE bitset_view(BitsetT* bitset_ptr, IdxT bitset_len)
+    : bitset_ptr_{bitset_ptr}, bitset_len_{bitset_len}
+  {
+  }
+  /**
+   * @brief Create a bitset view from a device vector view of the bitset.
+   *
+   * @param bitset_span Device vector view of the bitset
+   */
+  _RAFT_HOST_DEVICE bitset_view(raft::device_vector_view<BitsetT, IdxT> bitset_span)
+    : bitset_ptr_{bitset_span.data_handle()}, bitset_len_{bitset_span.extent(0)}
+  {
+  }
+  /**
+   * @brief Device function to test if a given index is set in the bitset.
+   *
+   * @param sample_index Single index to test
+   * @return bool True if index has not been unset in the bitset
+   */
+  inline _RAFT_DEVICE auto test(const IdxT sample_index) const -> bool
+  {
+    const IdxT bit_element = bitset_ptr_[sample_index / bitset_element_size];
+    const IdxT bit_index   = sample_index % bitset_element_size;
+    const bool is_bit_set  = (bit_element & (1ULL << bit_index)) != 0;
+    return is_bit_set;
+  }
+  /**
+   * @brief Get the device pointer to the bitset.
+   */
+  inline _RAFT_HOST_DEVICE auto get_bitset_ptr() -> BitsetT* { return bitset_ptr_; }
+  inline _RAFT_HOST_DEVICE auto get_bitset_ptr() const -> const BitsetT* { return bitset_ptr_; }
+  /**
+   * @brief Get the length of the bitset representation.
+   */
+  inline _RAFT_HOST_DEVICE auto get_bitset_len() const -> IdxT { return bitset_len_; }
+
+ private:
+  BitsetT* bitset_ptr_;
+  IdxT bitset_len_;
+};
+
+/**
+ * @brief RAFT Bitset.
+ *
+ * This structure encapsulates a bitset in device memory. It provides a view() method to get a
+ * device-usable lightweight view of the bitset.
+ * Each index is represented by a single bit in the bitset. The total number of bytes used is
+ * ceil(bitset_len / 4).
+ * The underlying type of the bitset array is uint32_t.
+ * @tparam IdxT Indexing type used. Default is uint32_t.
+ */
+template <typename IdxT = uint32_t>
+struct bitset {
+  using BitsetT            = uint32_t;
+  IdxT bitset_element_size = sizeof(BitsetT) * 8;
+
+  /**
+   * @brief Construct a new bitset object with a list of indices to unset.
+   *
+   * @param res RAFT resources
+   * @param mask_index List of indices to unset in the bitset
+   * @param bitset_len Length of the bitset
+   */
+  bitset(const raft::resources& res,
+         raft::device_vector_view<const IdxT, IdxT> mask_index,
+         IdxT bitset_len)
+    : bitset_{raft::make_device_vector<BitsetT, IdxT>(
+        res, raft::ceildiv(bitset_len, bitset_element_size))}
+  {
+    RAFT_EXPECTS(mask_index.extent(0) <= bitset_len, "Mask index cannot be larger than bitset len");
+    cudaMemsetAsync(bitset_.data_handle(),
```

Comment:
Would it be more efficient to construct the data on host and then just transfer it all at once, rather than setting and unsetting the data like this? It would come at the cost of host memory footprint, but I suspect that it would be worth it for large bitsets.